Add documentation - Main Concepts - API (#1855)
kuzzleio/kuzzle: doc/2/api/api-protocols/3-mqtt/index.md

---
code: false
type: page
title: MQTT
description: MQTT protocol usage and configuration  
order: 300
---

# MQTT

MQTT is a lightweight messaging protocol designed for IoT devices and optimized for **high-latency** or **unreliable networks**.  

> It establish a **bi-directional communication channel** between a client and a server.

Reviewer: ```suggestion
It establishes a **bidirectional communication channel** between a client and a server.
```

```js
const mqtt = require('mqtt'),
  client = mqtt.connect({ host: 'localhost' });

// Sending a volatile message
client.publish(
  'Kuzzle/request',
  JSON.stringify({
    index: 'index',
    collection: 'collection',
    controller: 'realtime',
    action: 'publish',
    requestId: 'some unique ID',
    body: { some: 'message' }
  })
);

// Getting Kuzzle's response
client.on('message', (topic, raw) => {
  const message = JSON.parse(Buffer.from(raw));

  // API results topic
  if (topic === 'Kuzzle/response') {
    // Response to our "publish" request
    if (message.requestId === 'some unique ID') {
      console.log('Message publication result: ', message);
    }
  }
});
```

### Using Kuzzle subscriptions

Kuzzle allows to [subscribe](/core/2/api/controllers/realtime/subscribe) to messages and events using advanced filters.

Each time a subscription is sent, a dedicated MQTT topic is created, named after the `channel` property issued by Kuzzle.

Here are the steps to perform a Kuzzle subscription using MQTT:

- Send a subscription request to Kuzzle

> - Listen to the request's response to ge the `channel` identifier

Reviewer: ```suggestion
- Listen to the request's response to get the `channel` identifier
```

- Subscribe to the MQTT topic named after this channel identifier

Example using [MQTT Node module](https://www.npmjs.com/package/mqtt):

```js

> const mqtt = require('mqtt'),
>   client = mqtt.connect({ host: 'localhost' }),
>   channels = [];

Reviewer: ```suggestion
const mqtt = require('mqtt');

const client = mqtt.connect({ host: 'localhost' });
const channels = [];
```

// Sending a volatile message
client.publish(
  'Kuzzle/request',
  JSON.stringify({
    index: 'index',
    collection: 'collection',
    controller: 'realtime',
    action: 'subscribe',
    requestId: 'some unique ID',
    body: {
      term: {
        some: 'filter'
      }
    }
  })
);

// Getting Kuzzle's response
client.on('message', (topic, raw) => {
  const message = JSON.parse(Buffer.from(raw));

  // API results topic
  if (topic === 'Kuzzle/response') {
    // Response to our "publish" request
    if (message.requestId === 'some unique ID') {
      channels.push(message.result.channel);
      client.subscribe(message.result.channel);
    }
  } else if (channels.indexOf(topic) !== -1) {
    // Subscription notification
    console.log('Notification: ', message);
  }
});
```

## Authorizations

### Publishing

If `allowPubSub` is set to `false`, clients can only publish to the `requestTopic` topic (defaults to `Kuzzle/request`).

If `allowPubSub` is set to `true`, clients are only forbidden to publish to the `responseTopic` topic (defaults to `Kuzzle/response`).

:::warning
Wildcards subcriptions are not allowed
:::

> If a client tries to publich to an unauthorized topic, his connection will immediately be shut down by the server.

Reviewer: ```suggestion
If a client tries to publish to an unauthorized topic, his connection will immediately be shut down by the server.
```

    "protocols": {
      "mqtt": {
        // Set to true to enable MQTT support
        "enabled": false,

        // Allow MQTT pub/sub capabilities or restrict to Kuzzle requests only
        "allowPubSub": false,
        
        // Switches responseTopic back to a regular public topic
        "developmentMode": false,

        // Delay in ms to apply between a disconnection notification is
        // received and the connection is actually removed
        "disconnectDelay": 250,

        // Name of the topic listened by the plugin for requests
        "requestTopic": "Kuzzle/request",

        // Name of the topic clients should listen to get requests result
        "responseTopic": "Kuzzle/response",

        // Constructor options passed to underlying mqtt server.
        // See aedes documentation for further reference: https://github.com/moscajs/aedes
        "server": {
          "port": 1883
        }
      }
```

## Usage

### Sending an API request and getting the response

By default, the MQTT protocol listens to the `Kuzzle/request` MQTT topic (see [configuration](#configuration)) for requests to the [Kuzzle API](/core/2/api/essentials/connecting-to-kuzzle).

> It then forwards Kuzzle response to the `Kuzzle/response` MQTT topic, **and only to the client who made the initial request**.

Reviewer 2: ```suggestion
It then forwards Kuzzle responses to the `Kuzzle/response` MQTT topic, **and only to the client who made the initial request**.
```

> The order of responses is not guaranteed to be the same than the order of requests. To link a response to its original request, use the `requestId` attribute: the response will have the same `requestId` than the one provided in the request.

Reviewer 2: ```suggestion
The order of responses is not guaranteed to be the same as the order of the requests. To link a response to its original request, use the `requestId` attribute: the response will have the same `requestId` than the one provided in the request.
```

const client = mqtt.connect({ host: 'localhost' });
const channels = [];

// Sending a volatile message
client.publish(
  'Kuzzle/request',
  JSON.stringify({
    index: 'index',
    collection: 'collection',
    controller: 'realtime',
    action: 'subscribe',
    requestId: 'some unique ID',
    body: {
      term: {
        some: 'filter'
      }
    }
  })
);

// Getting Kuzzle's response
client.on('message', (topic, raw) => {
  const message = JSON.parse(Buffer.from(raw));

  // API results topic
  if (topic === 'Kuzzle/response') {
    // Response to our "publish" request
    if (message.requestId === 'some unique ID') {
      channels.push(message.result.channel);
      client.subscribe(message.result.channel);
    }
  } else if (channels.indexOf(topic) !== -1) {
    // Subscription notification
    console.log('Notification: ', message);
  }
});
```

## Authorizations

### Publishing

If `allowPubSub` is set to `false`, clients can only publish to the `requestTopic` topic (defaults to `Kuzzle/request`).

If `allowPubSub` is set to `true`, clients are only forbidden to publish to the `responseTopic` topic (defaults to `Kuzzle/response`).

:::warning

> Wildcards subcriptions are not allowed

Reviewer 2: ```suggestion
Wildcard subcriptions are not allowed
```

:::

> If a client tries to publish to an unauthorized topic, his connection will immediately be shut down by the server.

Reviewer 2: ```suggestion
If a client tries to publish to an unauthorized topic, its connection will immediately be shut down by the server.
```

Reviewer: I didn't notice that one :sweat_smile: 

Singular their should be preferred:

```suggestion
If a client tries to publish to an unauthorized topic, their connection will immediately be shut down by the server.
```

### Subscribing

Subscription attempts to the `requestTopic` topic (defaults to `Kuzzle/request`) are ignored: client requests can only be listened by the MQTT server.

## Development mode

The MQTT `Kuzzle/response` topic is by default a special topic that acts as a private channel. Each client receives its own responses only, offering a simple first security layer.

> While this behavior is urgently recommended in production, it can bring a small drawback when testing and developing applications: it does not allow using most CLI tools.

Reviewer 2: ```suggestion
While this behavior is highly recommended in production, it can bring a small drawback when testing and developing applications: it does not allow using most CLI tools.
```
